File: Insert_in_middle_of_linked_list.py
class Node:
    def __init__(self,data):
        self.data=data
        self.next=None

class LinkedList:
    def __init__(self):
        self.head=None

    def insert_end(self,val):
        newnode=Node(val)
        if not self.head:self.head=newnode;return
        temp=self.head
        while temp.next:temp=temp.next
        temp.next=newnode

    def insert_middle(self,val):
        temp=self.head
        length=0
        while temp: length+=1; temp=temp.next
        pos=length//2
        newnode=Node(val)
        if pos==0: newnode.next=self.head; self.head=newnode; return
        temp=self.head
        for _ in range(pos-1): temp=temp.next
        newnode.next=temp.next
        temp.next=newnode

    def display(self):
        temp=self.head
        while temp: print(temp.data,end=" "); temp=temp.next
        print()

ll=LinkedList()
ll.insert_end(10); ll.insert_end(20); ll.insert_end(30)
ll.insert_middle(15)
ll.display()

File: Insert_in_middle_of_linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct Node{
    int data;
    struct Node* next;
};

struct Node* head=NULL;

void insert_end(int val){
    struct Node* newnode=malloc(sizeof(struct Node));
    newnode->data=val;newnode->next=NULL;
    if(!head){head=newnode;return;}
    struct Node* temp=head;
    while(temp->next)temp=temp->next;
    temp->next=newnode;
}

void insert_middle(int val){
    int len=0;
    struct Node* temp=head;
    while(temp){len++;temp=temp->next;}
    int pos=len/2;
    struct Node* newnode=malloc(sizeof(struct Node));
    newnode->data=val;
    temp=head;
    for(int i=0;i<pos-1;i++)temp=temp->next;
    newnode->next=temp->next;
    temp->next=newnode;
}

void display(){
    struct Node* temp=head;
    while(temp){printf("%d ",temp->data);temp=temp->next;}
    printf("\n");
}

int main(){
    insert_end(10);insert_end(20);insert_end(30);
    insert_middle(15);
    display();
    return 0;
}
